src/Components/Shared/src/WebRootComponentManager.cs

```diff
@@ -0,0 +1,175 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using static Microsoft.AspNetCore.Internal.LinkerFlags;
+
+#if COMPONENTS_SERVER
+namespace Microsoft.AspNetCore.Components.Server.Circuits;
+
+using Renderer = RemoteRenderer;
+
+internal partial class RemoteRenderer
+#elif COMPONENTS_WEBASSEMBLY
+namespace Microsoft.AspNetCore.Components.WebAssembly.Rendering;
+
+using Renderer = WebAssemblyRenderer;
+
+internal partial class WebAssemblyRenderer
+#else
+#error WebRootComponentManager cannot be defined in this assembly.
+#endif
+{
+    private WebRootComponentManager? _webRootComponentManager;
+
+    public WebRootComponentManager GetOrCreateWebRootComponentManager()
+        => _webRootComponentManager ??= new(this);
+
+    // Manages components that get added, updated, or removed in Blazor Web scenarios
+    // via Blazor endpoint invocations.
+    public sealed class WebRootComponentManager(Renderer renderer)
```

Comment:
I know this is a bit unusual. Normally, we would likely put this type of functionality in a base class, but we're trying to avoid changes to public API as much as possible. This class adds a `WebRootComponentManager` class to `RemoteRenderer` and `WebAssemblyRenderer` to manage components rendered by Blazor endpoints. It's a nested class so it can access protected members from each renderer. We also lazily instantiate the `WebRootComponentManager` instance for each renderer so that pure Server or WebAssembly scenarios don't have to pay the cost of the extra bookkeeping.

In the future, we may want to consider pulling this into public API so we don't have to share code in this manner.

Reply:
It's fine to share code in this manner if it's an implementation detail.